neuralpp/symbolic/sympy_expression.py

```diff
@@ -1,75 +1,121 @@
+from typing import List, Any, Type, Callable, Tuple, Optional, Dict
 from abc import ABC
 
 import sympy
 import operator
 import builtins
-from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant
+import fractions
+from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant, \
+    FunctionNotTypedError, NotTypedError, FunctionType
+from neuralpp.symbolic.basic_expression import BasicConstant, new_type, infer_python_callable_type
+from neuralpp.util.util import consistently_update_dict
+
 
 # In this file's doc, I try to avoid the term `sympy expression` because it could mean both sympy.Expr (or sympy.Basic)
 # and SymPyExpression. I usually use "sympy object" to refer to the former and "expression" to refer to the latter.
 
 
-def python_callable_to_sympy_function(python_callable: Callable) -> Type[sympy.Basic]:
-    match python_callable:
-        # boolean operation
-        case operator.__and__:
-            return sympy.And
-        case operator.__or__:
-            return sympy.Or
-        case operator.__not__:
-            return sympy.Not
-        case operator.__xor__:
-            return sympy.Xor
-        # comparison
-        case operator.le:
-            return sympy.Le
-        case operator.lt:
-            return sympy.Lt
-        case operator.ge:
-            return sympy.Ge
-        case operator.gt:
-            return sympy.Gt
-        case operator.eq:
-            return sympy.Eq
-        # arithmetic
-        case operator.add:
-            return sympy.Add
-        case operator.mul:
-            return sympy.Mul
-        case operator.pow:
-            return sympy.Pow
-        # min/max
-        case builtins.min:
-            return sympy.Min
-        case builtins.max:
-            return sympy.Max
+def infer_sympy_object_type(sympy_object: sympy.Basic, type_dict: Dict[sympy.Basic, Expression]) -> Expression:
+    match sympy_object:
+        case sympy.Integer():
+            return BasicConstant(int)
+        case sympy.Float():
+            return BasicConstant(float)
+        case sympy.Rational():
+            return BasicConstant(fractions.Fraction)
+        case sympy.logic.boolalg.BooleanAtom():
+            return BasicConstant(bool)
         case _:
-            raise ValueError(f"Python callable {python_callable} is not recognized.")
+            # It is obvious that we can look up type_dict for symbols like `symbol("x")`.
+            # We can also look up for function, such as `a+b`, where type_dict records the function type (instead of
+            # return type) of the (root-level) function.
+            try:
+                return type_dict[sympy_object]
+            except KeyError:  # if it's not in type_dict, try figure out ourselves
+                return BasicConstant(infer_python_callable_type(sympy_function_to_python_callable(sympy_object)))
+
+
+python_callable_and_sympy_function_pair = [
```

Comment:
Should the last word in the name be "pairs" rather than "pair"?

Reply:
Perhaps "relation" (as in math) would be even better.

Reply:
yes